SyncTests/BatchingClientTests.swift

```diff
@@ -0,0 +1,257 @@
+        return jsonFromRecord(record)?.toString()
+    }
+
+    private func mockRecord() -> Record<CleartextPayloadJSON> {
+        let jA = "{\"id\":\"mock\",\"histUri\":\"http://foo.com/\",\"title\": \"ñ\",\"visits\":[{\"date\":1222222222222222,\"type\":1}]}"
+        return Record<CleartextPayloadJSON>(id: "mock",payload: CleartextPayloadJSON(JSON.parse(jA)), modified: 10000, sortindex: 123, ttl: 1000000)
+    }
+
+    private func sizeOfMockRecord() -> ByteCount {
+        return serializeRecord(mockRecord())!.utf8.count
+    }
+
+    private func generateMockRecords(count: Int) -> [Record<CleartextPayloadJSON>] {
+        return (0..<count).reduce([]) { previous, _ in
+            return previous + [mockRecord()]
+        }
+    }
+
+    private func jsonFromRecord<T>(record: Record<T>) -> JSON? {
+        return JSON([
+            "id": record.id,
+            "foo": "bar"
+        ])
+    }
+
+    private func sizeReducer(total: ByteCount, record: Record<CleartextPayloadJSON>) -> ByteCount {
+        return total + (self.serializeRecord(record) ?? "").utf8.count
+    }
+
+    // Always return a gigantic encoded payload.
+    func massivify(record: Record<CleartextPayloadJSON>) -> JSON? {
+        return JSON([
+            "id": record.id,
+            "foo": String(count: Sync15StorageClient.maxRecordSizeBytes + 1, repeatedValue: "X" as Character)
+        ])
+    }
+}
+
+// MARK: Mock uploading tests
+extension Sync15BatchClientTests {
+
+    func testNoUploadWhenEmpty() {
+        var onCollectionUploadCalled = false
+        var onUploadCalled = false
+
+        let uploader: BatchUploadFunction = { _ in
+            onUploadCalled = true
+            return deferMaybe(self.emptyResponse)
+        }
+
+        let onCollectionUpload: (POSTResult -> Void) = { _ in onCollectionUploadCalled = true }
+
+        let batch = Sync15BatchClient(config: miniConfig, ifUnmodifiedSince: nil, serializeRecord: serializeRecord, uploader: uploader)
+        batch.addRecords([])
+        batch.commit(onCollectionUpload).succeeded()
+
+        // Shouldn't invoke the callback if we didn't actually upload anything
+        XCTAssertFalse(onCollectionUploadCalled)
+        XCTAssertFalse(onUploadCalled)
+    }
+
+    func testSinglePOSTUpload() {
+        var collectionUploadCount = 0
+        var uploadOpCount = 0
+
+        let uploader: BatchUploadFunction = { lines, timestamp, queryParams in
+            // Single POST should not have query parameters attached
+            XCTAssertNil(queryParams)
+            uploadOpCount += 1
+            return deferMaybe(self.emptyResponse)
+        }
+
+        let onCollectionUpload: (POSTResult -> Void) = { _ in collectionUploadCount += 1 }
+
+        let batch = Sync15BatchClient(config: miniConfig, ifUnmodifiedSince: nil, serializeRecord: serializeRecord, uploader: uploader)
+        let records = generateMockRecords(miniConfig.maxPostRecords)
```

Comment:
White-box testing! `maxPostRecords + 1` will verify whether my observation above is actually a bug.